pkg/handlers/cni/calico/handler.go

```diff
@@ -0,0 +1,310 @@
+	ctrl "sigs.k8s.io/controller-runtime"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	"github.com/d2iq-labs/capi-runtime-extensions/pkg/handlers"
+	"github.com/d2iq-labs/capi-runtime-extensions/pkg/handlers/cni"
+	"github.com/d2iq-labs/capi-runtime-extensions/pkg/k8s/client"
+	"github.com/d2iq-labs/capi-runtime-extensions/pkg/k8s/parser"
+)
+
+const (
+	CNILabelValue = "calico"
+
+	DefaultPodSubnet = "192.168.0.0/16"
+)
+
+type CalicoCNI struct {
+	client ctrlclient.Client
+}
+
+var (
+	_ handlers.NamedHandler                                 = &CalicoCNI{}
+	_ handlers.AfterControlPlaneInitializedLifecycleHandler = &CalicoCNI{}
+
+	//go:embed manifests/tigera-operator-configmap.yaml
+	tigeraConfigMapBytes []byte
+
+	// Only need to parse this on start-up and only once. If this isn't a valid configmap then this
+	// will panic on startup, and exit early.
+	tigeraConfigMap = parser.MustParseToObjects[*corev1.ConfigMap](tigeraConfigMapBytes)[0].(*corev1.ConfigMap)
+
+	//go:embed manifests/docker
+	dockerCNIManifests embed.FS
+
+	providerManifestsFS = map[string]embed.FS{
+		"DockerCluster": dockerCNIManifests,
+	}
+
+	calicoInstallationGK = schema.GroupKind{Group: "operator.tigera.io", Kind: "Installation"}
+)
+
+func New(c ctrlclient.Client) *CalicoCNI {
+	return &CalicoCNI{
+		client: c,
+	}
+}
+
+func (s *CalicoCNI) Name() string {
+	return "calico-cni"
+}
+
+func (s *CalicoCNI) AfterControlPlaneInitialized(
+	ctx context.Context,
+	req *runtimehooksv1.AfterControlPlaneInitializedRequest,
+	resp *runtimehooksv1.AfterControlPlaneInitializedResponse,
+) {
+	clusterKey := ctrlclient.ObjectKeyFromObject(&req.Cluster)
+
+	log := ctrl.LoggerFrom(ctx).WithValues(
+		"cluster",
+		clusterKey,
+	)
+
+	// Safe to set this to success response before we actually do the apply as the error handling
+	// will update for failure response properly.
+	resp.SetStatus(runtimehooksv1.ResponseStatusSuccess)
+
+	if v, ok := req.Cluster.GetLabels()[cni.CNILabelKey]; !ok || v != CNILabelValue {
+		log.V(4).Info("Skipping Calico CNI handler, CNI provider is not calico")
+		return
+	}
+
+	manifestsFS, ok := providerManifestsFS[req.Cluster.Spec.InfrastructureRef.Kind]
+	if !ok {
+		log.V(4).Info("Skipping Calico CNI handler, unknown CNI provider")
+		return
+	}
+
+	if err := applyTigeraOperatorCRS(ctx, s.client, &req.Cluster, log); err != nil {
+		log.Error(err, "failed to apply Tigera ClusterResourceSet")
+		resp.SetStatus(runtimehooksv1.ResponseStatusFailure)
+		resp.SetMessage(fmt.Sprintf("failed to apply Tigera ClusterResourceSet: %v", err))
+	}
+
+	if err := applyProviderCNICRS(ctx, manifestsFS, s.client, &req.Cluster, log); err != nil {
+		log.Error(err, "failed to apply CNI installation ClusterResourceSet")
+		resp.SetStatus(runtimehooksv1.ResponseStatusFailure)
+		resp.SetMessage(fmt.Sprintf("failed to apply CNI installation ClusterResourceSet: %v", err))
+	}
```

Comment:
Should this hook block until the installation is complete? Could we verify that CNI is working correctly?

Reply:
This is one of the problems of using CRS, there is nothing to wait on really other than whether the manifests are applied in remote cluster. I don't think there's a need to block until cni is running, rest of cluster deployment can continue while the CNI pods are started imo. We should be very careful about when we block.

Reply:
At most we could wait for the `ClusterResourceSetBinding`, but I think that would only to surface errors and set Conditions on the Cluster object.

Reply:
I think a condition to say CNI is deployed and working is a good idea, but what should we check for? I feel the same with CSI and CPI btw.